File: cli/src/main/java/io/microconfig/cli/commands/Command.java
package io.microconfig.cli.commands;

import io.microconfig.cli.CliException;
import io.microconfig.cli.CliFlags;
import io.microconfig.cli.ssl.TrustManagerFactory;

import javax.net.ssl.SSLContext;
import java.net.http.HttpRequest;

import static io.microconfig.cli.ssl.TrustManagerFactory.defaultTrust;
import static io.microconfig.cli.ssl.TrustManagerFactory.trustAll;

public abstract class Command {
    final CliFlags flags;
    final String[] args;

    public Command(String[] args) {
        this.args = args;
        this.flags = new CliFlags(args);
    }

    void addHeaders(HttpRequest.Builder request) {
        flags.type().ifPresent(t -> request.setHeader("X-TYPE", t));
        flags.branch().ifPresent(b -> request.setHeader("X-BRANCH", b));
        flags.tag().ifPresent(t -> request.setHeader("X-TAG", t));
        flags.vars().forEach((key, value) -> request.header("X-VAR", key + "=" + value));
    }

    String server() {
        return flags.server()
                .orElse(System.getenv().getOrDefault("MCS_ADDRESS", "http://localhost:8080"));
    }

    String component(String message) {
        if (args.length < 2) throw new CliException(message, 3);
        return args[1];
    }

    public abstract int execute();

    SSLContext sslContext() {
        if (flags.skipTls()) return trustAll();
        return flags.rootCa().map(TrustManagerFactory::rootCa).orElse(defaultTrust());
    }

}

File: cli/src/main/kotlin/io/microconfig/cli/commands/Command.kt
package io.microconfig.cli.commands

import io.microconfig.cli.CliException
import io.microconfig.cli.CliFlags
import io.microconfig.server.client.ClientConfig
import io.microconfig.server.client.MicroconfigClientImpl
import io.microconfig.server.client.MicroconfigException
import io.microconfig.server.client.MicroconfigRequest
import io.microconfig.server.client.ServiceConfig
import io.microconfig.server.common.HttpException
import io.microconfig.server.common.ssl.defaultTrust
import io.microconfig.server.common.ssl.rootCa
import io.microconfig.server.common.ssl.trustAll
import java.lang.System.getenv
import java.time.Duration
import javax.net.ssl.SSLContext

abstract class Command(val args: Array<String>) {
    val flags: CliFlags = CliFlags(args)
    private val client = MicroconfigClientImpl(ClientConfig(server(), timeout(), sslContext()))

    abstract fun execute(): Int

    fun component(message: String): String {
        if (args.size < 2) throw CliException(message, 3)
        return args[1]
    }

    fun configs(component: String, type: String? = null): List<ServiceConfig> {
        try {
            val request = configsRequest(component, type)
            return client.configs(request)
        } catch (e: MicroconfigException) {
            throw CliException(e.message!!, 200)
        } catch (e: HttpException) {
            throw CliException(e.message!!, 100)
        }
    }

    private fun configsRequest(component: String, type: String?): MicroconfigRequest {
        val env = flags.env() ?: "default"
        val ref = flags.branch() ?: flags.tag() ?: flags.ref()
        return MicroconfigRequest(component, env, type, ref, flags.vars())
    }

    private fun server(): String {
        return flags.server() ?: getenv()["MCS_ADDRESS"] ?: "http://localhost:8080"
    }

    private fun timeout(): Duration {
        return Duration.ofSeconds(flags.timeout())
    }

    private fun sslContext(): SSLContext {
        return if (flags.skipTls()) trustAll() else flags.rootCa()?.let { rootCa(it) } ?: defaultTrust()
    }

}